AVRO-3054: Fix / support decimal logical type in Ruby

andrewthauer opened the pull request:
### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3054) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3054
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - README/CHANGELOG not updated, but the logical type support from AVRO-2677 should have included this encoder/decoder rather than leaving it up to consumers to shove the correct bytes in there. AVRO-2677 is closed, and we are unsure the appropriate measure to take with that. Should we have opened a new issue due to AVRO-2677 not being fully implemented or reopened it? If the former, if someone wants to create that ticket, we're happy to rewrite our commit messages with the correct ticket number.

### Credit

This upstream PR was originally created by @ziggythehamster & @johvet as part of #918.

### Supercedes PRs (from $918)

The following PRs are currently open and implement an incorrect version of this feature, and should be closed:

- #829
- #840

Both PRs shove a string like "1.234" into the bytes, rather than encoding them according to the specification. Both PRs do not validate inputs nor introduce infrastructure to do that.

### Notes from #918

> The Avro specification is imprecise about how decimals are to be implemented, which required us to dig into the source code of Avro for Java as well as dig into Java's BigDecimal and BigInteger to make sure we were doing the same thing. Perhaps the specification could include a Java one-liner that implements the encoder/decoder? Here's a Scala one-liner that we used to test our implementation:

```java
val encoded = new java.math.BigDecimal("3.4562").setScale(6).unscaledValue().toByteArray()
val decoded = new java.math.BigDecimal(new java.math.BigInteger(encoded), 6)

encoded.map("%02x".format(_)).mkString(" ") // 34 bc c8: String
decoded // 3.456200: java.math.BigDecimal
```

> - We tested this in Ruby 2.3, 2.4, 2.5, 2.6, and 2.7. This is the reason for the <= check for retained objects, as some Ruby versions retain objects where others don't. We think this is either a bug in memory_profiler or a bug in Ruby itself.
> - Your build system depends on the echoe gem, but echoe is not compatible with RubyGems > 2.7. RubyGems 3.x has been out since 2018, and RubyGems 2.7 barely works in newer versions of Ruby. Consider upgrading this.
> - This PR is against master, but 1.9 is the current stable version. This branch is a version based on 1.9 with #761 incorporated (#761 was the PR that incompletely implemented AVRO-2677).
> - There's a gem published to GitHub packages as well, if you're like us and need a version with decimal support before this hits an official channel.

ziggythehamster commented:
This looks correct to me. Thanks for taking over on getting this landed; we have not had the time to work on the comments raised in the original PR or keep it fresh.

tjwp commented:
@andrewthauer Thank you for continuing to push this forward! I've taken a look at this branch, and I created my own branch to work through any additional changes that I would make. I have that commit here if you'd like to take a look: 

https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55

The main change is to bring the validation of the decimal logical type forward to when the schema is parsed. That way errors with a schema are caught prior to any encoding/decoding. Please take a look and let me know what you think.

As for https://github.com/apache/avro/pull/1082, if you and @ziggythehamster have looked and have no major concerns with it, then I'm happy to move forward.

andrewthauer commented:
@tjwp - Thanks for looking at this and also working on the additional validation refactoring.

It definitely seems more appropriate to handle the validation up front vs during encoding/decoding. https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55 seems cleaner as well imo. I haven't tested against that or #1082, but nothing stands out as concern in either. Are you thinking of getting both of those merged and before this?

andrewthauer commented:
@twjp - Did you want me to try and incorporate the changes from #1082 and/or https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55 into this PR?

tjwp commented:
@andrewthauer My plan was to address the schema matching with fixed feedback in #1082, and get that merged first. 

After that this pull request will need to be rebased. Then the changes from https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55 should be incorporated into this branch. It is possible some additional changes may also be required as a result of #1082. I may be able to work on this consolidation once #1082 is done.

andrewthauer commented:
@tjwp - Sounds good. I'll keep an eye on #1082, and once that lands I'll rebase and look at incorporating https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55.

andrewthauer commented:
@tjwp - I've rebased #1082 into this branch and re-tested. Everything seems good. Thanks!

I've added commit https://github.com/apache/avro/pull/1102/commits/fa8b7254e76f3af5b2ec7d2ebdfdfafb26fd5316 which avoids frozen string errors in tests since I'm testing on ruby 2.7.

Is there anything else you'd like addressed in this PR or does that mostly cover it?

andrewthauer commented:
Almost forgot about https://github.com/tjwp/avro/commit/bdb64e658750e58ecdc8e659195cb39df41dbb55. This is not incorporated into the branch @tjwp.

tjwp commented on lang/ruby/test/test_logical_types.rb:
```diff
@@ -99,6 +100,140 @@ def test_parse_fixed_duration
+      end
+
+      assert_equal 5, report.total_allocated
+      assert_equal 0, report.total_retained
```
This line is failing with Ruby 2.6 where it is getting 1. This should probably be changed to match https://github.com/apache/avro/pull/1102/files#diff-5c5f6ae9be9e29a9812a5b49746e79d50c326fb32fc984064dc15c52a7b41385R212-R213 including a similar comment.

andrewthauer replied:
I noticed this too, but I can't reproduce it locally (even with the same ruby 2.6.7). I didn't realize the details here. Happy to change that quick.

andrewthauer replied:
Updated based on suggestion

tjwp commented on lang/ruby/test/test_logical_types.rb:
```diff
@@ -99,6 +100,140 @@ def test_parse_fixed_duration
+      end
+
+      assert_equal 5, report.total_allocated
+      # Ruby 2.3, 2.4 and 2.7 do not retain anything. Ruby 2.5 and some 2.6 retain 1
```
Only Ruby 2.6 and later are supported now, so references to earlier versions could be dropped here. (optional)

andrewthauer replied:
Updated comment

tjwp commented on lang/ruby/lib/avro/logical_types.rb:
```diff
@@ -17,9 +17,193 @@
+      # @param schema [Avro::Schema]
+      #     The schema defining precision and scale for the conversion
+      #
+      # @raise [ArgumentError]
```
Minor, but this initialize method no longer raises since the validation was moved to the `BytesSchema` class.

Now that parsing and matching has been added for fixed decimal types the same validation should be performed there, and it would be nice to share the implementation. I think that can happen separately.

andrewthauer replied:
Removed yardoc comment

tjwp reviewed (commented):
@andrewthauer Thank you for sticking with this! 

There is a test that is failing after I enabled the github actions to run for this PR. Once that is fixed, then I will go ahead and merge this. 

There is some follow-up work related to this that would be great to have, including validating fixed decimals and supporting the encoding/decoding of fixed decimals, but I think those can happen separately.

andrewthauer commented:
@tjwp - Updated as per your suggestions. Ideally the additional validation can be addressed in future PR. Would be great to get this PR shipped so we can stop using the forked version we have.

tjwp commented:
I've created issues for the fixed validation and encoding/decoding:
- https://issues.apache.org/jira/browse/AVRO-3141
- https://issues.apache.org/jira/browse/AVRO-3142
I'm optimistic that I'll have some time to work on those.

tjwp commented:
>  Ideally the additional validation can be addressed in future PR.

👍🏻 

> Would be great to get this PR shipped so we can stop using the forked version we have.

Looks good! Merging now
